ibis/backends/pandas/udf.py

```diff
@@ -92,13 +92,20 @@ def execute_udf_node_groupby(op, *args, **kwargs):
     @execute_node.register(
         ops.ElementWiseVectorizedUDF, *(itertools.repeat(object, nargs))
     )
-    def execute_udf_node(op, *args, **kwargs):
+    def execute_udf_node(op, *args, cache=None, **kwargs):
         # We have rewritten op.func to be a closure enclosing
         # the kwargs, and therefore, we do not need to pass
         # kwargs here. This is true for all udf execution in this
         # file.
         # See ibis.udf.vectorized.UserDefinedFunction
-        return op.func(*args)
+
+        # prevent executing UDFs multiple times on different execution branches
+        try:
+            result = cache[op]
```

Comment:
I think you also want to pass the timecontext in here, otherwise the results might be wrong (same UDF subtree appears in both the left subtree and right subtree with different time context) and ideally have a test for this case (asof join the same table with UDF, right subtree time context would be expanded so you will have the same op but with different time context)

Reply:
@icexelloss is it possible that https://github.com/ibis-project/ibis/blob/master/ibis/backends/tests/test_timecontext.py#L107 already covers it?

Reply:
I think that test name might be confusing. It is a window operations but I don't think the context are adjusted because it's unbounded window (basically a groupby demean operation). I am not sure what's a good example to demonstrate the issue but I think since `Scope` uses op + timecontext as the lookup key, we should probably use the same key since the semantics of the cache here is the same as `Scope` object if I understand correctly. I am a concerned about subtle correctness bugs if we don't pass timecontext here.

Reply:
@icexelloss updated, now using a compound `(op, timecontext)` key. 

Could you please add a test case in a follow-up since you are more experienced with the timecontext handling?

Reply:
Sounds good. Not sure if I can tackle this soon but happy to leave as follow ups.